src/Microsoft.Health.Fhir.Shared.Core/Features/Validation/ProfileValidator.cs

```diff
@@ -4,43 +4,125 @@
+
+                var packagesToInclude = new HashSet<string>();
+
+                if (_coreConfig.PackageConfiguration.IncludeDefaultPackages)
+                {
+                    packagesToInclude.Add(packageVariables.CorePackageName);
+                    packagesToInclude.Add(packageVariables.ExpansionsPackageName);
+
+                    // elements package name is not included in the reflection metadata, but we should try and include it
+                    var extensionsPackageName = packageVariables.ExpansionsPackageName.Replace("expansions", "elements", StringComparison.OrdinalIgnoreCase);
+                    packagesToInclude.Add(extensionsPackageName);
+                }
+
+                foreach (var package in _coreConfig.PackageConfiguration.PackageNames)
+                {
+                    packagesToInclude.Add(package);
+                }
+
+                FhirPackageSource validationSource;
+
+                if (!string.IsNullOrEmpty(_coreConfig.PackageConfiguration.PackageSource) &&
+                    !_coreConfig.PackageConfiguration.PackageSource.StartsWith("http", StringComparison.OrdinalIgnoreCase))
+                {
+                    // Allow local packages
+                    var resolvedPath = Path.GetFullPath(_coreConfig.PackageConfiguration.PackageSource, Environment.CurrentDirectory);
+                    validationSource = new FhirPackageSource(Directory.GetFiles(resolvedPath)
+                        .Where(f => packagesToInclude.Any(p => f.Contains(p, StringComparison.OrdinalIgnoreCase)
+                        && f.EndsWith("tgz", StringComparison.OrdinalIgnoreCase))).ToArray());
+                }
+                else
+                {
+                    // Look for packages on an npm server
+                    var server = string.IsNullOrEmpty(_coreConfig.PackageConfiguration.PackageSource) ? packageVariables.Server : _coreConfig.PackageConfiguration.PackageSource;
+                    validationSource = new FhirPackageSource(server, packagesToInclude.ToArray());
+                }
+
+                // Ensure packages are downloaded to temp folder
+                var patient = validationSource.ListResourceUris(ResourceType.Patient).ToArray();
+
+                return new MultiResolver(_profilesResolver, new CachedResolver(validationSource, _options.CacheDurationInSeconds));
             }
-            catch (Exception)
+            catch (Exception ex)
             {
-                // Something went wrong during profile loading, what should we do?
+                _logger.LogError(ex, "Error initializing profile validator");
                 throw;
             }
         }
 
+        internal static (string Server, string CorePackageName, string ExpansionsPackageName) GetFhirPackageVariables()
+        {
+            Type type = typeof(FhirPackageSource);
```

Comment:
Where is this coming from? I can't find reference to this FhirPackageSource or where its fields are defined.

Reply:
These are compiled into the Firely libraries, it isn't a great dependency, but its a way to find the NPM urls. Perhaps we should define these ourselves?
There is a unit test to cover they exist: https://github.com/microsoft/fhir-server/pull/3241/files#diff-98805c9eb649d4044301a52d69bb13b1e09768a441ded93ce851277986c8103aR18